api/routes.go

```diff
@@ -89,11 +92,30 @@ func (api *API) noAuth(h HandlerWithUser) httprouter.Handle {
 }
 
 // withAuth ensures that the user making the request has logged in.
-func (api *API) withAuth(h HandlerWithUser) httprouter.Handle {
+func (api *API) withAuth(h HandlerWithUser, allowsAPIKey bool) httprouter.Handle {
 	return func(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
 		api.logRequest(req)
+
+		// Check for a token.
+		u, token, err := api.userAndTokenByRequestToken(req)
+		if err == nil {
+			// Embed the verified token in the context of the request.
+			ctx := jwt.ContextWithToken(req.Context(), token)
+			h(u, w, req.WithContext(ctx), ps)
+			return
+		}
+
 		// Check for an API key.
-		u, token, err := api.userAndTokenByAPIKey(req)
+		ak, err := apiKeyFromRequest(req)
```

Comment:
f/u suggestion

`apiKeyFromRequest` should only be called in this method. Then it attaches the api key to the context.
All handlers read the token from the context.

That's for extra safety in case of potential inconsistencies because we are only checking whether API Key authentication is allowed in this method and not in the handlers.

Reply:
This is almost exactly what's happening. The only exceptions are the two handlers which don't use the `withAuth` middleware because they need to allow unauthenticated requests as well - `userLimitsGET` and `userLimitsSkylinkGET`.

Outside of those two, we only check the API key here, then fetch the issuing user and pass it to the handlers directly.